state/processor.go

```diff
@@ -101,15 +101,20 @@ func (tp *TransactionProcessor) addStateToHistory(layer types.LayerID, newHash c
 
 }
 
-func (tp *TransactionProcessor) ApplyRewards(layer types.LayerID, miners map[string]struct{}, underQuota map[string]struct{}, bonusReward, diminishedReward *big.Int) {
-	for minerId, _ := range miners {
-		if _, ok := underQuota[minerId]; ok {
-			tp.globalState.AddBalance(address.HexToAddress(minerId), diminishedReward)
+func (tp *TransactionProcessor) ApplyRewards(layer types.LayerID, miners []string, underQuota map[string]int, bonusReward, diminishedReward *big.Int) {
+	for _, minerId := range miners {
+		//if we have 2 blocks in same layer, one of them can receive a diminished reward and the other cannot
+		if val, ok := underQuota[minerId]; ok {
+			if val > 0 {
+				tp.globalState.AddBalance(address.HexToAddress(minerId), diminishedReward)
+				underQuota[minerId] = underQuota[minerId] - 1
+			} else {
+				tp.globalState.AddBalance(address.HexToAddress(minerId), bonusReward)
```

Comment:
you cloud delete the map entry when `underQuota[minerId]` goes down to 0 and get rid of this `else and line, right ?

Reply:
because then ok will be false after it gone done to 0